contracts/hydro/src/testing.rs

```diff
@@ -3763,3 +3292,218 @@ fn test_cannot_vote_while_long_deployment_ongoing() {
+        Coin::new(2000u128, IBC_DENOM_2),
+        Coin::new(4000u128, ST_ATOM_ON_NEUTRON),
+    ] {
+        let locking_info = get_message_info(&deps.api, user_address, &[token_to_lock]);
+        let msg = ExecuteMsg::LockTokens {
+            lock_duration: init_msg.lock_epoch_length * 2,
+        };
+
+        let res = execute(deps.as_mut(), env.clone(), locking_info.clone(), msg);
+        assert!(res.is_ok());
+    }
+
+    // Have user vote for both proposals with all of its lockups
+    let voting_info = get_message_info(&deps.api, user_address, &[]);
+    let msg = ExecuteMsg::Vote {
+        tranche_id,
+        proposals_votes: vec![
+            ProposalToLockups {
+                proposal_id: first_proposal_id,
+                lock_ids: vec![0, 1, 2],
+            },
+            ProposalToLockups {
+                proposal_id: second_proposal_id,
+                lock_ids: vec![3, 4, 5],
+            },
+        ],
+    };
+
+    let res = execute(deps.as_mut(), env.clone(), voting_info.clone(), msg);
+    assert!(res.is_ok());
+
+    verify_proposals_and_rounds_powers(
+        &deps.storage,
+        first_round_id,
+        tranche_id,
+        &[(first_proposal_id, 11875), (second_proposal_id, 11250)],
+        &[(first_round_id, 23125), (second_round_id, 18500)],
+    );
+
+    // Mock that stATOM token group ratio changed to 1.6
+    setup_st_atom_token_info_provider_mock(
+        &mut deps,
+        token_info_provider_addr.clone(),
+        Decimal::from_str("1.6").unwrap(),
+    );
+
+    // Update stATOM token ratio, which should update all proposal powers and round total powers
+    let msg = ExecuteMsg::UpdateTokenGroupRatio {
+        token_group_id: ST_ATOM_TOKEN_GROUP.to_string(),
+        old_ratio: Decimal::from_str("1.5").unwrap(),
+        new_ratio: Decimal::from_str("1.6").unwrap(),
+    };
+
+    let info = get_message_info(&deps.api, token_info_provider, &[]);
+    let res = execute(deps.as_mut(), env.clone(), info, msg);
+    assert!(res.is_ok());
+
+    verify_proposals_and_rounds_powers(
+        &deps.storage,
+        first_round_id,
+        tranche_id,
+        &[(first_proposal_id, 12000), (second_proposal_id, 11750)],
+        &[(first_round_id, 23750), (second_round_id, 19000)],
+    );
+
+    // Remove both stATOM and LSM token info provider. This should bring all proposal and round powers to zero
+    for token_provider in [
+        token_info_provider_addr.to_string(),
+        LSM_TOKEN_INFO_PROVIDER_ID.to_string(),
+    ] {
+        let msg = ExecuteMsg::RemoveTokenInfoProvider {
+            provider_id: token_provider,
+        };
+
+        let res = execute(deps.as_mut(), env.clone(), admin_info.clone(), msg);
+        assert!(res.is_ok());
+    }
+
+    verify_proposals_and_rounds_powers(
+        &deps.storage,
+        first_round_id,
+        tranche_id,
+        &[(first_proposal_id, 0), (second_proposal_id, 0)],
+        &[(first_round_id, 0), (second_round_id, 0)],
+    );
```

Comment:
I think we should test somewhere that its possible to unlock tokens after the respective token_info_provider was removed

Reply:
commented here because it seems like the easiest place to check that

Reply:
> I think we should test somewhere that its possible to unlock tokens after the respective token_info_provider was removed

We don't allow this currently, even before token info providers were introduced. It's the same with LSTs, e.g. if some validator drops out of the top N, users still wouldn't be able to unlock their tokens until their lockup expires.
I agree we should add this functionality. But also check if user voted for some longer proposal deployment and don't allow unlock in that case. Since it is not entirely related to this PR, I would do this in a separate one. Would you agree?

Reply:
Just to clarify: I am not talking about immediately unlocking. I am just talking about being able to unlock at all when the lockup expires

Reply:
We don't have problem with that, since we don't do denom validation on token unlocks. Just collect all the expired ones and send them back to the user.